tools/codegen/dest/lazy_ir.py

```diff
@@ -0,0 +1,387 @@
+                                     scalarT, scalarTypeT)
+import tools.codegen.api.meta as meta
+import tools.codegen.api.cpp as cpp
+import tools.codegen.api.structured as structured
+from tools.codegen.api.translate import translate
+from tools.codegen.selective_build.selector import SelectiveBuilder
+
+valueListT = BaseCppType('std', 'vector<ir::Value>')
+valueT = BaseCppType('ir', 'Value')
+intArrayT = BaseCppType('std', 'vector<int64_t>')
+
+def process_ir_type(typ: BaseTy, mutable: bool, binds: str) -> NamedCType:
+    """
+    This function takes a type from NativeFunctions and converts it for use with
+    lazy tensor codegen.  Currently its output is used in several places, and so far
+    it has been possible for them to all use the same conversions, but that may not be
+    optimal or possible in the finished system.
+
+    Type conversion for lazy currently consists of
+     (1) changing Tensor-like things into Value-like things
+     (2) wrapping everythign in a BaseCType
+     (3) wrapping everything further in a NamedCType
+
+    (1) converts Tensors to Values since Values are how Lazy IR represents tensors.  There
+    is special handling for Optional[Tensor] or List[Tensor], etc- hence 'tensor-like'
+
+    This is incomplete- there are assertions in places that it's expected to need to add
+    more types as the codegen is used with more operators.
+    """
+    if isinstance(typ, BaseType):
+        if typ.name == BaseTy.Tensor:
+            return NamedCType(binds, BaseCType(valueT))
+        elif typ.name == BaseTy.Scalar:
+            return NamedCType(binds, BaseCType(scalarT))
+        else:
+            raise AssertionError(f"TODO add support for type {repr(typ)}")
+    elif isinstance(typ, OptionalType):
+        if str(typ.elem) == 'Tensor':
+            return NamedCType(binds, OptionalCType(BaseCType(valueT)))
+        elif str(typ.elem) == 'ScalarType':
+            return NamedCType(binds, OptionalCType(BaseCType(scalarTypeT)))
+        else:
+            raise AssertionError(f"TODO add support for type {repr(typ)}")
+    elif isinstance(typ, ListType):
+        if str(typ.elem) == 'Tensor':
+            return NamedCType(binds, BaseCType(valueListT))
+        elif str(typ.elem) == 'Tensor?':
+            return NamedCType(binds, ListCType(OptionalCType(BaseCType(valueT))))
+        elif str(typ.elem) == 'int':
+            return NamedCType(binds, BaseCType(intArrayT))
+        else:
+            raise AssertionError(f"TODO add support for type {repr(typ)}")
+    else:
+        raise AssertionError(f"unrecognized type {repr(typ)}")
+
+
+def isValueType(typ: NamedCType) -> bool:
+    """
+    Given a type, determine if it is a Value-like type.  This is equivalent to
+    being Tensor-like, but assumes the type has already been transformed.
+    """
+    if isinstance(typ.type, BaseCType):
+        return typ.type.type.ns == valueT.ns and typ.type.type.name == valueT.name
```

Comment:
does `return type.type.type == valueT` not work?

You'll also have to a case for `isinstance(typ.type, VectorCType)`